drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c

```diff
@@ -357,7 +357,7 @@ static int mipi_dbi_flexio_lcdif_reset(const struct device *dev, uint32_t delay)
 		return err;
 	}
 
-	k_msleep(delay);
+	k_sleep(delay);
```

Comment:
can you help me to understand where the int64 division happens with the k_msleep? because, it is just an inline wrapper to k_sleep...

Reply:
Sure. The issue is in the milliseconds to ticks conversion:

```c
static inline int32_t k_msleep(int32_t ms)
{
	return k_sleep(Z_TIMEOUT_MS(ms));
}
```

where Z_TIMEOUT_MS is:

```
# define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS((k_ticks_t)k_ms_to_ticks_ceil64(MAX(t, 0)))
```

which is

```
#define k_ms_to_ticks_ceil64(t) 	z_tmcvt_64(t, Z_HZ_ms, Z_HZ_ticks, true, true, false)
```

which resolves to

```
		 z_tmcvt_int_div_64(__t, __from_hz, __to_hz, __round_up, __round_off) \
```

which is

```
	(((uint64_t) (__t) + z_tmcvt_off_div(__from_hz, __to_hz,
					     __round_up, __round_off)) /z_tmcvt_divisor(__from_hz, __to_hz))
```

(phew!)

This scales using a uint64_t multiple and divide combo at runtime.

Reply:
The same applies for `K_MSEC()`, isn't it?

> #define K_MSEC(ms)     Z_TIMEOUT_MS(ms)

`z_tmcvt_64` doesn't resolve always to `z_tmcvt_int_div_64()`, does it? It depends on macros `Z_HZ_ms` and `Z_HZ_ticks`.

Reply:
Sure but the constant value is known at compile time so the compiler can inline it. Tell me more?

Reply:
Sorry, I still don't understand  how `k_msleep()` leads to a int64_t division but `K_MSEC()` doesn't.

Reply:
I think I'm missing something here as well- the implementation of `k_msleep` looks like so:

```
static inline int32_t k_msleep(int32_t ms)                                                                  
{                                                                                                           
         return k_sleep(Z_TIMEOUT_MS(ms));                                                                   
}
```

and the `mipi_dbi_reset` function (as implemented in this PR) calls `K_MSEC()`, which expands to:

```
#define K_MSEC(ms)     Z_TIMEOUT_MS(ms)
```

is there some reason that these would end up being appreciably different? It seems like both will need to perform the same conversion and link in the 64 bit division functions

Reply:
Let me try another aproach, see if that helps.

Let's take `display_st7735r.c` as an example. Before this change, the important parts are:

```c
#define ST7735R_RESET_TIME      1
...
static int st7735r_reset_display(const struct device *dev)
{
...	ret = mipi_dbi_reset(config->mipi_dev, ST7735R_RESET_TIME);

static inline int mipi_dbi_reset(const struct device *dev, uint32_t delay_ms)
{
...
	return api->reset(dev, delay_ms);

static int mipi_dbi_spi_reset(const struct device *dev, uint32_t delay_ms)
{
...
	k_msleep(delay_ms);  // AKA ksleep(K_MSEC(delay_ms)); 
```

The compiler can't tell anything about `delay_ms` so `K_MSEC` has to the multiply/divide at runtime.

Compare with after:

```c
#define ST7735R_RESET_TIME      1
...
static int st7735r_reset_display(const struct device *dev)
{
...	ret = mipi_dbi_reset(config->mipi_dev, ST7735R_RESET_TIME);

static inline int mipi_dbi_reset(const struct device *dev, uint32_t delay_ms)
{
...
	return api->reset(dev, K_MSEC(delay_ms));
```

As `ST7735R_RESET_TIME` is constant, and `mipi_dbi_reset` is inline, then the compiler can constant fold the K_MSEC and do the multipy/divide at compile time.

Reply:
Yup, thank you for explaining it again :) Seems like the key change here is that the `mipi_dbi_reset` portion will be inlined, so the compiler can take advantage of optimizations here that aren't available in the driver implementation (due to the fact that function is hidden behind a function pointer).